assets/src/higher-order-components/with-latest-checkin.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * External imports
+ */
+import { withSelect, withDispatch } from '@wordpress/data';
+import { createHigherOrderComponent, compose } from '@wordpress/compose';
+import { isModelEntityOfModel } from '@eventespresso/validators';
+
+/**
+ * Exposes properties related to the latest check-in entity for a given
+ * registration and datetime.
+ *
+ * Properties passed through to wrapped component are:
+ *
+ * - checkInEntity {BaseEntity|null} The check-in entity that is related to the
+ * given registration and datetime id.
+ * - hasResolvedCheckin {boolean} Whether the check-in entity selector has
+ * resolved.  This is important because its possible there is no entity for
+ * this registration and datetime if that registration hasn never been checked
+ * in.
+ * - onClick {function} A click handler which when invoked, will toggle the
+ * check-in state for the given registration and datetimeId. Note: this will
+ * replace the store latestCheckin record in the state for this given
+ * registration and datetime id which will get picked up by the `withSelect`
+ * HOC in the composed component.
+ *
+ * @type {WPComponent}
+ */
+const withLatestCheckin = createHigherOrderComponent(
+	compose( [
+		withSelect(
+			( select, { registration, datetimeId } ) => {
+				if ( ! isModelEntityOfModel(
+					registration,
+					'registration'
+				) ) {
+					return {};
+				}
+				const { getLatestCheckin } = select( 'eventespresso/core' );
+				const { hasFinishedResolution } = select( 'core/data' );
+				return {
+					checkinEntity: getLatestCheckin( registration.id, datetimeId ),
+					hasResolvedCheckin: hasFinishedResolution(
```

Comment:
won't this increase the number of renders in wrapped components because it will be returning every result that gets yielded by the generators?

In the admin refactor stuff I have primarily been using `hasFinishedResolution()` to determine whether to return the final results as opposed to the default "loading" results.

For example I would have done something like the following with this:

```jsx
const checkinEntity = getLatestCheckin( registration.id, datetimeId );
const hasResolvedCheckin = hasFinishedResolution(
	'eventespresso/core',
	'getLatestCheckin',
	[ registration.id, datetimeId ]
);
return hasResolvedCheckin ?
	{ checkinEntity, loadingCheckin: false } :
	{ checkinEntity: null, loadingCheckin: true};
```

which means that the wrapped component only sees its props change when `hasFinishedResolution()` returns `true`.

Is this not a good way to do things?

What are the benefits to how you are doing things?

Reply:
Well I did it my way predominately because I figured that provides some flexibility for the implementing component to decide things but I think your approach is better because it _ensures_ consuming components don't implement things in a way that are less performant.  So ya good point and I'll implement something like what you suggested.

Reply:
So what I ended up implementing in 7cdf751 was simply normalizing an `undefined` result to `null` for the `checkinEntity` property since that's most likely what you see in debugging.  When a selector with a resolver is resolving the initial returned value from the select call is `undefined`.  The return value when there is no checkIn record is `null`, so I normalized things so that `null` is swapped in for `undefined` to keep renders to a minimum.

I didn't change the `hasResolvedCheckin` prop because essentially that will either be true or false anyways, so splitting it out into a separate variable accomplishes nothing different.